File: src/bootstrap_omp.c
#include <gsl/gsl_rng.h>
#include <gsl/gsl_randist.h>
#include <gsl/gsl_statistics.h>
#include <math.h>
#include <time.h>
#include <sys/time.h>
#include <omp.h>
#include "tools.h"

void bootstrap_multidim (double x[], double y[], double* result, int* b, int* B, int *n, int* d);
void samp_k_from_n (int* k, int* n, int* a, gsl_rng *restrict r);

void bootstrap_b(double x[], double *result, int * b, int * B, int * n, gsl_rng *restrict r)
{
  //printf("\nI am now here.\n\n");
  double T_boot[*B]; // to store result of doing statistic (mean) on each replicate
  
  //int a[n];
  
  for (int i = 0; i < *B; i++) {
    
    //printf("\n i = %d\n", i);
    //samp_k_from_n(n, n, a, r);
    //printf("B\n\n");
    
    double x_star[*n];
    gsl_ran_sample(r, x_star, *n, x, *b, sizeof(double));
    
    T_boot[i] = gsl_stats_mean(x_star, 1, *n);
    //printf("\n\n T_[%i] = %f\n\n", i, T_boot[i]);
  }
  
  
  //printf("\n\n I've finished the for loop\n\n");
  //double temp = sqrt(gsl_stats_variance(T_boot, 1, B));
  //printf("\n\ntemp is %f\n\n", temp);
  
  *result = sqrt(gsl_stats_variance(T_boot, 1, *B));
  //printf("\n\nresult is %f\n\n", *result);
  //printf("\n\nresult is %f\n\n", sqrt(gsl_stats_variance(T_boot, 1, B)));
  
}


void bootstrap_b_omp(double x[], double *result, int *b, int *B, int *n)
{
  //printf("\nI am now here.\n\n");
  // double T_boot[*B]; // to store result of doing statistic (mean) on each replicate
  
  double theta_star = 0.0;
  double theta_star_sq = 0.0;
  
  #pragma omp parallel reduction(+:theta_star, theta_star_sq)
  {
    gsl_rng *restrict r = NULL;
    if(r == NULL) { // First call to this function, setup RNG
      gsl_rng_env_setup();
      r = gsl_rng_alloc(gsl_rng_mt19937);
    }
    gsl_rng_set(r, time(NULL) + omp_get_thread_num() * 1024);
    
    #pragma omp for 
    for (int i = 0; i < *B; i++) {
    
      double theta_star_temp;
      double theta_star_sq_temp;
    
      double x_star[*n];
      gsl_ran_sample(r, x_star, *n, x, *b, sizeof(double));
    
      theta_star_temp = gsl_stats_mean(x_star, 1, *n);
      theta_star_sq_temp = pow(theta_star_temp, 2);
      theta_star += theta_star_temp;
      theta_star_sq += theta_star_sq_temp;
    //printf("\n\n T_[%i] = %f\n\n", i, T_boot[i]);
    }
  }
  double theta_star_bar = theta_star / (double) *B;
  double var;
  var = (theta_star_sq - *B * pow(theta_star_bar, 2)) / (*B - 1);
  *result = sqrt(var);
}

void BLB_serial(double x[], double *result, double *gamma, int *s, int *R, int *n)
  // n is length of the data
{
  static gsl_rng *restrict r = NULL;
  
  if(r == NULL) { // First call to this function, setup RNG
    gsl_rng_env_setup();
    r = gsl_rng_alloc(gsl_rng_mt19937);
    gsl_rng_set(r, time(NULL));
  }
  
  int b;
  b = (int) floor(pow(*n, *gamma));
  
  double xis[*s];
  
  for (int i = 0; i < *s; i++)
  {
    // select subsample of size b of the data
    double subsamp[b];
    gsl_ran_choose(r, subsamp, b, x, *n, sizeof(double));
    
    // run bootstrap on that subsample (resamples n for each replicate)
    bootstrap_b(subsamp, &xis[i], &b, R, n, r);
  }
  
  // take average of results from the s subsamples
  *result = gsl_stats_mean(xis, 1, *s);
}

void BLB_omp_on_s(double x[], double *result, double *gamma, int *s, int *R, int *n)
  // n is length of the data 
{
  int b;
  b = (int) floor(pow(*n, *gamma));
  
  double xi = 0;
  
  #pragma omp parallel reduction(+:xi)
  {
    gsl_rng *restrict r = NULL;
    if(r == NULL) { // First call to this function, setup RNG
      gsl_rng_env_setup();
      r = gsl_rng_alloc(gsl_rng_mt19937);
    }
    gsl_rng_set(r, time(NULL) + omp_get_thread_num() * 1024);
    
    #pragma omp for
    for (int i = 0; i < *s; i++)
    {
      double xi_temp;
      
      // select subsample of size b of the data
      double subsamp[b];
      gsl_ran_choose(r, subsamp, b, x, *n, sizeof(double));
      
      // run bootstrap on that subsample (resamples n for each replicate)
      bootstrap_b(subsamp, &xi_temp, &b, R, n, r);
      xi += xi_temp;
    }
  }
  // take average of results from the s subsamples
  xi = xi / (double) *s;
  *result = xi;	
}

/*
void BLB_md_omp_on_s(double x[], double y[], double *result, double *gamma, int *s, int *R, int *n, int *d)
  // n is length of the data 
{
  int b;
  b = (int) floor(pow(*n, *gamma));
  
  double xi[*d];
  
#pragma omp parallel reduction(+:xi)
{
  gsl_rng *restrict r = NULL;
  if(r == NULL) { // First call to this function, setup RNG
    gsl_rng_env_setup();
    r = gsl_rng_alloc(gsl_rng_mt19937);
  }
  gsl_rng_set(r, time(NULL) + omp_get_thread_num() * 1024);
  
#pragma omp for
  for (int i = 0; i < *s; i++)
  {
    double xi_temp[*d];
    
    // select subsample of size b of the data
    int subsamp_idx[b];
    samp_k_from_n(b, n, subsamp_idx, r);
    double subsamp_x[b * *d];
    
    double subsamp_y[b];
    for (int j = 0; j < b; j++) {
      subsamp_y[j] = y[subsamp_idx[j]];
      for (int i = 0; i < *d; i++) {
        subsamp_x[i + j * *d] = x[i + subsamp_idx[j] * *d];
      }
     }
      
    // run bootstrap on that subsample (resamples n for each replicate)
    bootstrap_multidim(subsamp_x, subsamp_y, xi_temp, &b, R, n, d);
    
    for (int j = 0; j < *d; j++) {
      xi[j] += xi_temp[j];
    }
  }
}
// take average of results from the s subsamples

  for (int j = 0; j < *d; j++) {
    xi[j] = xi[j] / (double) *s;
  }
  result = xi;	
}
*/

void BLB_omp_on_B(double x[], double *result, double *gamma, int *s, int *R, int *n)
  // n is length of the data
{
  gsl_rng *restrict r = NULL;
  
  if(r == NULL) { // First call to this function, setup RNG
    gsl_rng_env_setup();
    r = gsl_rng_alloc(gsl_rng_mt19937);
    gsl_rng_set(r, time(NULL));
  }
  
  int b;
  b = (int) floor(pow(*n, *gamma));
  
  double xis[*s];
  
  for (int i = 0; i < *s; i++)
  {
    // select subsample of size b of the data
    double subsamp[b];
    gsl_ran_choose(r, subsamp, b, x, *n, sizeof(double));
    
    // run bootstrap on that subsample (resamples n for each replicate)
    bootstrap_b_omp(subsamp, &xis[i], &b, R, n);
  }
  
  // take average of results from the s subsamples
  *result = gsl_stats_mean(xis, 1, *s);
}

void BLB_omp_on_sB(double x[], double *result, double *gamma, int *s, int *R, int *n)
  // n is length of the data 
{
  int b;
  b = (int) floor(pow(*n, *gamma));
  
  double xi = 0;
  
#pragma omp parallel reduction(+:xi)
{
  gsl_rng *restrict r = NULL;
  if(r == NULL) { // First call to this function, setup RNG
    gsl_rng_env_setup();
    r = gsl_rng_alloc(gsl_rng_mt19937);
  }
  gsl_rng_set(r, time(NULL) + omp_get_thread_num() * 1024);
  
#pragma omp for
  for (int i = 0; i < *s; i++)
  {
    double xi_temp;
    
    // select subsample of size b of the data
    double subsamp[b];
    gsl_ran_choose(r, subsamp, b, x, *n, sizeof(double));
    
    // run bootstrap on that subsample (resamples n for each replicate)
    bootstrap_b_omp(subsamp, &xi_temp, &b, R, n);
    xi += xi_temp;
  }
}
// take average of results from the s subsamples
xi = xi / (double) *s;
*result = xi;	
}


void bootstrap_for_clust(double x[], double *result, double *gamma, int *R, int *n, int *i)
  // n is length of the data
{
  gsl_rng *restrict r = NULL;
  
  if(r == NULL) { // First call to this function, setup RNG
    gsl_rng_env_setup();
    r = gsl_rng_alloc(gsl_rng_mt19937);
    gsl_rng_set(r, time(NULL)+*i*1024);
  }
  
  int b;
  b = (int) floor(pow(*n, *gamma));
  
  double xi = 0;
  
  // select subsample of size b of the data
  double subsamp[b];
  gsl_ran_choose(r, subsamp, b, x, *n, sizeof(double));
    
  // run bootstrap on that subsample (resamples n for each replicate)
  // runs in parallel through openMP
  bootstrap_b_omp(subsamp, &xi, &b, R, n);
  
  *result = xi;
}

void bootstrap_b_multinom(double subsamp[], double *result, int *b, int *B, int *n, gsl_rng *restrict r)
{
  //printf("\nI am now here.\n\n");
  double theta_star[*B]; // to store result of doing statistic (mean) on each replicate
  
  for (int i = 0; i < *B; i++) {
    
    // indices for resample
    unsigned int w[*b]; // where the weights (# of each index) will be stored
    double p[*b]; //probs for multinomial resampling, then set equal
    for (int j = 0; j < *b; j++) {
      p[j] = 1;
    }
    gsl_ran_multinomial(r, *b, *n, p, w);
    
    double v[*b];
    for (int j = 0; j < *b; j++) {
      v[j] = (double) w[j];
      
      //printf("W[%i] = %i\n", j, w[j]);
      //printf("V[%i] = %f\n", j, v[j]);
      //printf("subsamp[%i] = %f\n", j, subsamp[j]);
    }
    //printf("\n");
    
    // calculate theta_star
    theta_star[i] = gsl_stats_wmean(v, 1, subsamp, 1, *b);
    //printf("%f\n", theta_star[i]);
    //printf("\n\n T_[%i] = %f\n\n", i, theta_star[i]);
  }
  
  
  //printf("\n\n I've finished the for loop\n\n");
  //double temp = sqrt(gsl_stats_variance(T_boot, 1, B));
  //printf("\n\ntemp is %f\n\n", temp);
  
  *result = sqrt(gsl_stats_variance(theta_star, 1, *B));
  //printf("\n\nresult is %f\n\n", *result);
  //printf("\n\nresult is %f\n\n", sqrt(gsl_stats_variance(T_boot, 1, B)));
}  

void bootstrap_b_omp_multinom(double subsamp[], double *result, int *b, int *B, int *n)
{
  //printf("\nI am now here.\n\n");
  // double T_boot[*B]; // to store result of doing statistic (mean) on each replicate
  
  double theta_star = 0.0;
  double theta_star_sq = 0.0;
  
#pragma omp parallel reduction(+:theta_star, theta_star_sq)
{
  gsl_rng *restrict r = NULL;
  if(r == NULL) { // First call to this function, setup RNG
    gsl_rng_env_setup();
    r = gsl_rng_alloc(gsl_rng_mt19937);
  }
  gsl_rng_set(r, time(NULL) + omp_get_thread_num() * 1024);
  
#pragma omp for 
  for (int i = 0; i < *B; i++) {
    
    double theta_star_temp;
    double theta_star_sq_temp;
    
    // indices for resample
    unsigned int w[*b]; // where the weights (# of each index) will be stored
    double p[*b]; //probs for multinomial resampling, then set equal
    for (int j = 0; j < *b; j++) {
      p[j] = 1;
    }
    gsl_ran_multinomial(r, *b, *n, p, w);
    
    double v[*b];
    for (int j = 0; j < *b; j++) {
      v[j] = (double) w[j];
      
      //printf("W[%i] = %i\n", j, w[j]);
      //printf("V[%i] = %f\n", j, v[j]);
      //printf("subsamp[%i] = %f\n", j, subsamp[j]);
    }
    //printf("\n");
    
    // calculate theta_star
    theta_star_temp = gsl_stats_wmean(v, 1, subsamp, 1, *b);
    theta_star_sq_temp = pow(theta_star_temp, 2);
    theta_star += theta_star_temp;
    theta_star_sq += theta_star_sq_temp;
    //printf("\n\n T_[%i] = %f\n\n", i, T_boot[i]);
  }
}
double theta_star_bar = theta_star / (double) *B;
double var;
var = (theta_star_sq - *B * pow(theta_star_bar, 2)) / (*B - 1);
*result = sqrt(var);
}


void BLB_serial_multinom(double x[], double *result, double *gamma, int *s, int *R, int *n)
  // n is length of the data
{
  static gsl_rng *restrict r = NULL;
  
  if(r == NULL) { // First call to this function, setup RNG
    gsl_rng_env_setup();
    r = gsl_rng_alloc(gsl_rng_mt19937);
    gsl_rng_set(r, time(NULL));
  }
  
  int b;
  b = (int) floor(pow(*n, *gamma));
  
  double xis[*s];
  
  for (int i = 0; i < *s; i++)
  {
    // select subsample of size b of the data
    double subsamp[b];
    gsl_ran_choose(r, subsamp, b, x, *n, sizeof(double));
    
    // run bootstrap on that subsample (uses weights)
    bootstrap_b_multinom(subsamp, &xis[i], &b, R, n, r);
  }
  
  // take average of results from the s subsamples
  *result = gsl_stats_mean(xis, 1, *s);
}

void BLB_omp_on_s_multinom(double x[], double *result, double *gamma, int *s, int *R, int *n)
  // n is length of the data 
{
  int b;
  b = (int) floor(pow(*n, *gamma));
  
  double xi = 0;
  
#pragma omp parallel reduction(+:xi)
{
  gsl_rng *restrict r = NULL;
  if(r == NULL) { // First call to this function, setup RNG
    gsl_rng_env_setup();
    r = gsl_rng_alloc(gsl_rng_mt19937);
  }
  gsl_rng_set(r, time(NULL) + omp_get_thread_num() * 1024);
  
#pragma omp for
  for (int i = 0; i < *s; i++)
  {
    double xi_temp;
    
    // select subsample of size b of the data
    double subsamp[b];
    gsl_ran_choose(r, subsamp, b, x, *n, sizeof(double));
    
    // run bootstrap on that subsample (resamples n for each replicate)
    bootstrap_b_multinom(subsamp, &xi_temp, &b, R, n, r);
    xi += xi_temp;
  }
}
// take average of results from the s subsamples
xi = xi / (double) *s;
*result = xi;	
}

void BLB_omp_on_B_multinom(double x[], double *result, double *gamma, int *s, int *R, int *n)
  // n is length of the data
{
  gsl_rng *restrict r = NULL;
  
  if(r == NULL) { // First call to this function, setup RNG
    gsl_rng_env_setup();
    r = gsl_rng_alloc(gsl_rng_mt19937);
    gsl_rng_set(r, time(NULL));
  }
  
  int b;
  b = (int) floor(pow(*n, *gamma));
  
  double xis[*s];
  
  for (int i = 0; i < *s; i++)
  {
    // select subsample of size b of the data
    double subsamp[b];
    gsl_ran_choose(r, subsamp, b, x, *n, sizeof(double));
    
    // run bootstrap on that subsample (resamples n for each replicate)
    bootstrap_b_omp_multinom(subsamp, &xis[i], &b, R, n);
  }
  
  // take average of results from the s subsamples
  *result = gsl_stats_mean(xis, 1, *s);
}

void BLB_omp_on_sB_multinom(double x[], double *result, double *gamma, int *s, int *R, int *n)
  // n is length of the data 
{
  int b;
  b = (int) floor(pow(*n, *gamma));
  
  double xi = 0;
  
#pragma omp parallel reduction(+:xi)
{
  gsl_rng *restrict r = NULL;
  if(r == NULL) { // First call to this function, setup RNG
    gsl_rng_env_setup();
    r = gsl_rng_alloc(gsl_rng_mt19937);
  }
  gsl_rng_set(r, time(NULL) + omp_get_thread_num() * 1024);
  
#pragma omp for
  for (int i = 0; i < *s; i++)
  {
    double xi_temp;
    
    // select subsample of size b of the data
    double subsamp[b];
    gsl_ran_choose(r, subsamp, b, x, *n, sizeof(double));
    
    // run bootstrap on that subsample (resamples n for each replicate)
    bootstrap_b_omp_multinom(subsamp, &xi_temp, &b, R, n);
    xi += xi_temp;
  }
}
// take average of results from the s subsamples
xi = xi / (double) *s;
*result = xi;	
}

void bootstrap_for_clust_multinom(double x[], double *result, double *gamma, int *R, int *n, int *i)
  // n is length of the data
{
  gsl_rng *restrict r = NULL;
  
  if(r == NULL) { // First call to this function, setup RNG
    gsl_rng_env_setup();
    r = gsl_rng_alloc(gsl_rng_mt19937);
    gsl_rng_set(r, time(NULL)+*i*1024);
  }
  
  int b;
  b = (int) floor(pow(*n, *gamma));
  
  double xi = 0;
  
  // select subsample of size b of the data
  double subsamp[b];
  gsl_ran_choose(r, subsamp, b, x, *n, sizeof(double));
  
  // run bootstrap on that subsample (resamples n for each replicate)
  // runs in parallel through openMP
  bootstrap_b_omp_multinom(subsamp, &xi, &b, R, n);
  
  *result = xi;
}

File: R/bootstrap_omp.R
#' Bag of Little Bootstraps for a vector
#' 
#' Implements Bag of Little Bootstraps on a 1-d data vector. The statistic which is bootstrapped is the mean, with the exception of BLB_R, which can accept a generic function as the statistic.
#' 
#' The functions described here implement the Bag of Little Bootstraps (BLB) with varying degrees of efficiency.
#' 
#' \code{BLB_R} implements BLB entirely in R. It is the most generic function, in that it can accept any function (which acts on a vector) as the statistic to be bootstrapped. It is also by far the slowest implementations.
#' 
#' The rest of the functions are all implemented in C, but are built in to only be able to bootstrap the mean.
#' 
#' \code{BLB_serial} does not take advantage of any parallelism. 
#' 
#' \code{BLB_omp_on_s} uses openMP to parallelise the for loop over the s subsampling steps, then runs a serial bootstrap function within each subsample.
#' 
#' \code{BLB_omp_on_r} does not parallelise over the s subsamples, but uses openMP to parallelise the for loop over the r resampling steps within each subsample.
#' 
#' \code{BLB_omp_on_rs} uses openMP to parallelise over both the subsampling and resampling loops.
#' 
#' \code{BLB_cluster} uses parallelism in R to send each of the s subsamples to different servers in the cluster, then uses openMP to parallelise the interations over the r resamples within each subsample.
#' 
#' @name BLB
#' @aliases BLB_serial
#' @aliases BLB_omp_on_s
#' @aliases BLB_omp_on_r
#' @aliases BLB_omp_on_sr
#' @aliases BLB_cluster
#' @param data A vector of the original sample
#' @param gamma Controls the size of the subsamples - each subsample is of size b = (floor(length(data) ^ gamma). Ideally in gamma is in [0.5, 1]
#' @param FUN A function to calculate the estimator of the parameter of interest
#' @param s The number of subsamples
#' @param r The number of bootstrap replications (resamples) per subsample
#' @return The mean across the s subsamples of the standard errors of the parameter estimates
#' @examples
#' X <- rnorm(1000000, 0, 5)
#' BLB_R(X, mean, gamma=0.6)
BLB_R <- function(data, gamma, FUN, ..., s=15, r=100) {
  n <- length(data)
  b <- round(n^gamma)
  xis <- numeric(s)
  
  subsamp_mat <- matrix(0, s, b)
  subsamp <- numeric(b)
  resample_mat <- matrix(0, r, n)
  
  for (i in 1:s) {
    subsamp <- sample(data, b)
    resample_mat <- matrix(sample(subsamp, r*n, replace = TRUE), r, n)
    theta <- apply(resample_mat, 1, FUN, ...)
    xis[i] <- sd(theta)
  }
  return(mean(xis))
}

#' @rdname BLB
#' @examples 
#' BLB_serial(X, 0.9)
#' @export
BLB_serial <- function(data, gamma, s = 15, r = 100) {
  ans <- .C("BLB_serial", as.double(data), result = as.double(0), as.double(gamma), as.integer(s), as.integer(r), as.integer(length(data)))
  return(ans$result)
}
#' @rdname BLB
#' @examples 
#' BLB_omp_on_s(X, 0.6)
#' @export
BLB_omp_on_s <- function(data, gamma, s = 15, r = 100) {
  ans <- .C("BLB_omp_on_s", as.double(data), result = as.double(0), as.double(gamma), as.integer(s), as.integer(r), as.integer(length(data)))
  return(ans$result)
}

#' @rdname BLB
#' @examples 
#' BLB_omp_on_r(X, 0.6)
#' @export
BLB_omp_on_r <- function(data, gamma, s = 15, r = 100) {
  ans <- .C("BLB_omp_on_B", as.double(data), result = as.double(0), as.double(gamma), as.integer(s), as.integer(r), as.integer(length(data)))
  return(ans$result)
}

#' @rdname BLB
#' @examples 
#' BLB_omp_on_sr(X, 0.6)
#' @export
BLB_omp_on_sr <- function(data, gamma, s = 15, r = 100) {
  ans <- .C("BLB_omp_on_sB", as.double(data), result = as.double(0), as.double(gamma), as.integer(s), as.integer(r), as.integer(length(data)))
  return(ans$result)
}


# void bootstrap_for_clust(double x[], double *result, double *gamma, int *R, int *n)
bootstrap_for_clust <- function(i, data, gamma, r = 100) {
  # a is a dummy first argument, as can use with clusterApply
  ans <- .C("bootstrap_for_clust", as.double(data), result = as.double(0), as.double(gamma), as.integer(r), as.integer(length(data)), as.integer(i))
  return(ans$result)
}

#' @rdname BLB
#' @examples 
#' cluster <- c("greywagtail", "greyheron", "greypartridge", "greyplover")
#' BLB_cluster(X, 0.6, cluster, s=16, r=96)
#' @export
BLB_cluster <- function(data, gamma, cluster, s=15, r=100) {
  clust <-makePSOCKcluster(cluster)
  clusterEvalQ(clust, library("BLB", lib.loc = "~/R"))
  lambda <- clusterApply(clust, 1:s, "bootstrap_for_clust", data=data, gamma=gamma, r=r)
  stopCluster(clust)  
  return(mean(unlist(lambda)))
}

# void BLB_md_omp_on_s(double x[], double y[], double *result, double *gamma, int *s, int *R, int *n, int *d)
#BLB_md_omp_on_s <- function(X, y, gamma, s=15, r=100) {
#  n <- nrow(X)
#  d <- ncol(X)
#  b <- floor(n^gamma)
#  ans <- .C("BLB_md_omp_on_s", as.double(t(X)), as.double(y), result = as.double(numeric(d)), as.double(gamma), as.integer(s), as.integer(r), as.integer(n), as.integer(d))
#  return(ans$result)
#}

# void BLB_serial_multinom(double x[], double *result, double *gamma, int *s, int *R, int *n)
BLB_serial_multinom <- function(data, gamma, s = 15, r = 100) {
  ans <- .C("BLB_serial_multinom", as.double(data), result = as.double(0), as.double(gamma), as.integer(s), as.integer(r), as.integer(length(data)))
  return(ans$result)
}

BLB_omp_on_s_multinom <- function(data, gamma, s = 15, r = 100) {
  ans <- .C("BLB_omp_on_s_multinom", as.double(data), result = as.double(0), as.double(gamma), as.integer(s), as.integer(r), as.integer(length(data)))
  return(ans$result)
}

BLB_omp_on_r_multinom <- function(data, gamma, s = 15, r = 100) {
  ans <- .C("BLB_omp_on_B_multinom", as.double(data), result = as.double(0), as.double(gamma), as.integer(s), as.integer(r), as.integer(length(data)))
  return(ans$result)
}

BLB_omp_on_sr_multinom <- function(data, gamma, s = 15, r = 100) {
  ans <- .C("BLB_omp_on_sB_multinom", as.double(data), result = as.double(0), as.double(gamma), as.integer(s), as.integer(r), as.integer(length(data)))
  return(ans$result)
}

bootstrap_for_clust_multinom <- function(i, data, gamma, r = 100) {
  # a is a dummy first argument, as can use with clusterApply
  ans <- .C("bootstrap_for_clust_multinom", as.double(data), result = as.double(0), as.double(gamma), as.integer(r), as.integer(length(data)), as.integer(i))
  return(ans$result)
}


BLB_cluster_multinom <- function(data, gamma, cluster, s=15, r=100) {
  clust <-makePSOCKcluster(cluster)
  clusterEvalQ(clust, library("BLB", lib.loc = "~/R"))
  lambda <- clusterApply(clust, 1:s, "bootstrap_for_clust_multinom", data=data, gamma=gamma, r=r)
  stopCluster(clust)  
  return(mean(unlist(lambda)))
}




